chain/sync.go

```diff
@@ -1404,6 +1423,22 @@ func (syncer *Syncer) syncFork(ctx context.Context, incoming *types.TipSet, know
 		}
 
 		if nts.Equals(tips[cur]) {
+			// We've identified the ancestor
+			// Still need to make sure this wouldn't cause us to fork away from the checkpointed tipset
+
+			chkpt := syncer.GetCheckpoint()
+
+			if chkpt != types.EmptyTSK {
+				chkTs, err := syncer.ChainStore().LoadTipSet(chkpt)
+				if err != nil {
+					return nil, xerrors.Errorf("failed to retrieve checkpoint tipset: %w", err)
+				}
+
+				if chkTs.Height() > nts.Height() {
```

Comment:
nit: i would also explicitly check that we're leaving the checkpoint tipset behind for more robustness; ideally comparing heights reflects this but there could be some corner cases (e.g., direct manipulation of chain by plumbing commands) that could be caught with that

Reply:
Yeah, we definitely should be checking specific hashes here.

Reply:
One potentially bad edge case here is if we mark a block as our checkpoint, but are currently synced to a different chain. This would prevent us from switching to that checkpoint.

Reply:
> One potentially bad edge case here is if we mark a block as our checkpoint, but are currently synced to a different chain.
That case is explicitly checked by the syncer when setting a checkpoint (that's why I marked this only as a nit, but still nice to have as an integrity check).

Reply:
changed to explicitly check the tipsetkeys we're forking away from

@whyrusleeping As Schomatis said, the potentially bad edge case should never happen, because we only allow you to mark tipsets currently in your heaviest chain as checkpointed -- we said that was okay for v0 of this, can you confirm it's still okay?

(pretty easy to change that, we can simply call SetHead when setting a checkpoint, but it's risky)

Reply:
@arajasek Thats sufficient for a first release of this, I guess the flow would be:
run a checkpoint import, then mark the checkpoint head as your checkpoint.